KAFKA-2440; Use `NetworkClient` instead of `SimpleConsumer` to fetch data from replica

ijuma opened the pull request.

ijuma committed:
Use `NetworkClient` in `ReplicaFetcherThread`

`ConsumerFetcherThread` still uses `SimpleConsumer`.

ijuma committed:
Call `metadataUpdater.maybeHandleDisconnection` from `NetworkClient.close()`

ijuma commented:
Tests passed locally. Will also run system tests.

ijuma commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -124,4 +176,81 @@ class ReplicaFetcherThread(name:String,
+    }
+  }
+
+  private def sendRequest(apiKey: ApiKeys, apiVersion: Option[Short], request: AbstractRequest): ClientResponse = {
```
I noticed that the equivalent method in `SimpleConsumer` retries once if there is a failure during connect, send or receive. On second failure, the exception is then propagated. Should we replicate that behaviour here?

junrao replied:
We don't need to replicate the reconnect behavior in SimpleConsumer. However, we probably need to close the client connection on socketTimeoutException.

ijuma replied:
Good point. I pushed a change where the connection is closed if any `Throwable` is thrown. Let me know if you think it's better to just close it on `SocketTimeoutException`.

ijuma commented:
System tests passed http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/72/

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -128,64 +114,64 @@ abstract class AbstractFetcherThread(name: String,
+                case _ =>
+                  if (isRunning.get) {
+                    error("Error for partition [%s,%d] to broker %d:%s".format(topic, partitionId, sourceBroker.id,
+                      partitionData.exception))
```
partitionData.exception is an Option. So, we probably should call get()?

ijuma replied:
Well spotted, fixed.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -17,47 +17,96 @@
+
+  override def shutdown(): Unit = {
+    super.shutdown()
+    networkClient.close()
```
This means that the selector to the same broker can be created multiple times over time. When this happens, the same sensors have already been registered in Metrics and will be registered again. This is probably fine since the existing sensor will just be used. Could you do a bit test to see re-creating the networkClient to the same broker is fine?

ijuma replied:
Good point. I think this won't work in that case, an exception will be thrown during the construction of the `Selector` instance. The following code is where the exception will eventually be thrown from, I think:

```
    synchronized void registerMetric(KafkaMetric metric) {
        MetricName metricName = metric.metricName();
        if (this.metrics.containsKey(metricName))
            throw new IllegalArgumentException("A metric named '" + metricName + "' already exists, can't register another one.");
        this.metrics.put(metricName, metric);
        for (MetricsReporter reporter : reporters)
            reporter.metricChange(metric);
    }
```

We ran into the same issue in the controller code. The solution we implemented for that case was to keep a map from broker id to `NetworkClient` in `KafkaController`. We could do something similar here (the map could live in `ReplicaManager`), but I find this quite inconvenient.

Do you think it would make sense to try to change this behaviour so that this exception is not thrown or perhaps allow it to be configurable in some way?

ijuma committed:
Add missing `get` after `partitionData.exception`

Bug spotted by Jun.

ijuma committed:
Improve error handling in `ReplicaFetcherThread.sendRequest`

Closing client after exceptions as suggested by Jun and improved
error message if `blockingSendAndReceive` times out.

ijuma committed:
Introduce `SslReplicaFetchTest`

And refactor `ReplicaFetchTest` into `BaseReplicaFetchTest` and
`PlaintextReplicaFetchTest`.

ijuma committed:
Introduce `SslTopicMetadataTest`

And refactor `TopicMetadataTest` into `BaseTopicMetadataTest` and
`PlaintextTopicMetadataTest`.

ijuma commented:
@junrao, I added a couple of tests and addressed your feedback with the exception of the metrics point where I left a question.

ijuma committed:
Unregister metrics and sensors on `Selector.close`

In order to do this, we had to implement removal of sensors and
metrics in `Metrics`.

ijuma committed:
Remove `KafkaController.networkClientMap`

We no longer need to do this because the metrics and sensors
are removed on `Selector.close`.

ijuma commented:
@junrao I added a couple of additional commits related to unregistering of metrics and sensors. Can you please take a look and let me know what you think?

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2440-use-network-client-in-fetcher

* apache/trunk:
  KAFKA-1893: Allow regex subscriptions in the new consumer
  KAFKA-2489: add benchmark for new consumer
  TRIVIAL: change to API doc for seekToEnd() to clarify lazy evaluation
  KAFKA-2453: Enable new consumer in EndToEndLatency

ijuma commented:
System tests passed:

http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/73/

auradkar commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -78,6 +78,28 @@ public void metricChange(KafkaMetric metric) {
+                if (mbean.metrics.isEmpty())
+                    unregister(mbean);
+                else
+                    reregister(mbean);
```
Never mind. I think I understand.

Old comment: "Can you explain the reason behind reregistering? On metric removal, you should simply unregister the Mbean right?"

auradkar commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -136,11 +138,39 @@ public synchronized Sensor sensor(String name, MetricConfig config, Sensor... pa
+            List<Sensor> childSensors = childrenSensors.get(name);
+            if (childSensors != null) {
+                for (Sensor childSensor : childSensors)
+                    removeSensor(childSensor.name());
```
Where do you cleanup the childrenSensors object? Otherwise we will maintain a reference to the Sensor objects always.

ijuma replied:
Good catch. `childrenSensors.get(name)` should be `childrenSensors.remove(name)`. Fixed locally and will push after running the tests.

auradkar commented:
Thanks for the patch. Couple of general comments about the metrics changes:
1. This works fine when we remove a Sensor using the Metrics reference. However, it is possible to use Sensors this way:

Sensor s;
void create() {
s = metrics.sensor(xx); 
}

void process {
  s.record()
}

Even if the metric is deleted, it is possible to continually record() changes and the writer thread will never know. Should s.record() throw an Exception? 
1. I think it is possible to deadlock if we call Metrics.removeSensor() and Sensor.add(MetricName..) at the same time.

ijuma committed:
Fix `childrenSensors` leak

Spotted by Aditya.

auradkar commented:
Overall, I'd like to use the metrics changes in KAFKA-2419. For now, I'll apply these changes locally to proceed with development.

ijuma commented:
Thanks for reviewing the changes @auradkar. About your points:
1. Interesting point about `s.record()` throwing an exception in this case. I'm not really sure to be honest. Do you think this would help in diagnosing cases where we remove a metric even though we continue to use it? Also, do you think this case could perhaps be considered as part of KAFKA-2419 instead of here?
2. Can you please elaborate on how you think it can deadlock?

auradkar commented:
1. I think it is totally fine to tackle this in KAFKA-2419 as opposed to here. I'm not sure if throwing an Exception is the right solution, but I think it is possible to use Sensor objects after they have been deleted especially when we create Sensors for entities that may be deleted and recreated. For example, Topic and client-id sensors
2. In general, they seem to call each other and acquire exclusive locks.
   For example: Sensor.add() acquires an exclusive lock on the sensor object and calls Metrics.registerMetric which acquires a lock on the Metrics object.  Separately, Metrics.removeSensor() acquires a lock on the Metrics object and calls Sensor.metrics() which locks the Sensor object

auradkar commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -190,6 +236,10 @@ synchronized void registerMetric(KafkaMetric metric) {
     }
 
+    ConcurrentMap<String, List<Sensor>> childrenSensors() {
+        return childrenSensors;
```
Minor comment. Perhaps wrap this in an unmodifiableMap

ijuma replied:
Will do. I'll also add a comment that this method should only be used for tests (ideally we'd have an annotation for that).

ijuma commented:
Thanks for explaining the potential deadlock @auradkar. Do you think it would be feasible to retrieve `Sensor.metrics` without locking? This would require changing the underlying list in the `Sensor` instance to be thread-safe (perhaps a `CopyOnWriteArrayList`).

If this is not desirable for some reason, then we may be able to use finer-grained locking in `removeSensor`, but the code would be more complex.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -557,6 +558,8 @@ private void addToCompletedReceives() {
         public final Sensor bytesReceived;
         public final Sensor selectTime;
         public final Sensor ioTime;
+        private final List<MetricName> topLevelMetricNames = new ArrayList<>();
```
Perhaps add a comment that this is for metrics that are not registered through sensors.

ijuma replied:
Good point, will do.

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:
```diff
@@ -149,6 +151,62 @@ public void testBadSensorHiearchy() {
+
+        metrics.removeSensor("test.parent1");
+        assertNull(metrics.getSensor("test.parent1"));
+        assertNull(metrics.metrics().get("test.parent1.count"));
```
We should pass in MetricName to get() instead of String, right?

ijuma replied:
Yes, you're right. Will fix this and other cases like it. Really annoying that this is not a compiler error in Java.

junrao commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -17,19 +17,20 @@
+import org.apache.kafka.common.protocol.Errors
+import AbstractFetcherThread._
+import scala.collection.{JavaConverters, mutable, Set, Map}
+import JavaConverters._
```
This seems unused.

ijuma replied:
Will remove.

aprilzt replied:
I found that you put forward the bug kafka-1711, where should i put the file? what's the direction. I hava download the file ConsoleProducer.scala,and also change the file accordint to the order, 2 line deleted and 4 line added.

ijuma committed:
Various minor fixes and improvements

Suggested by Jun and Aditya.

ijuma committed:
Tweak locking in `removeSensor` to avoid potential deadlock

This issue was noticed by Aditya.

ijuma commented:
Addressed reviews by @junrao and @auradkar. Please let me know if you have any other feedback.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -136,11 +139,55 @@ public synchronized Sensor sensor(String name, MetricConfig config, Sensor... pa
+            for (KafkaMetric metric : sensor.metrics())
+                removeMetric(metric.metricName());
+    }
+
```
Does that cause issue when a sensor/metric is added concurrently during removal? For example, 
1. removeSensor(n1): complete until line 173.
2. a new sensor is added and we add a metric of the same name (as the metrics to be removed in step 1).
3. removeSensor(n1): complete the rest of the steps.

After step 3, we may have removed the metrics added in step 2. Or step 2 will fail when adding the metric.

ijuma replied:
Step 2 will fail during `MetricName` registration as it has not yet been removed, yes. But this is not exactly worse than what we have in trunk now (where we basically end up throwing an exception after adding a sensor in a very similar way when a Selector is created a second time).

I think the concurrency strategy needs to be revisited within the context of removals and metrics/sensors collection, but I was hoping that could be done as part of KAFKA-2419. What do you think?

auradkar replied:
Perhaps we can lock the Sensor first and then lock the base metrics instance whenever we are removing a sensor. During addMetric, the sensor is locked first and then the Metrics object. If we follow the same pattern when doing removeSensor, we should not have any deadlocks right?

What do you think? I'm still open to handling this in KAFKA-2419, but I want to be sure we have an acceptable solution.

ijuma replied:
That may work. I need to go through the code again to double-check. @junrao, would you be happy with this solution instead? If so, I can do it before we merge as it's not a big change.

ijuma replied:
@auradkar, I pushed a commit implementing something like what you suggested here. Can you please check if this is OK in your mind?

ijuma committed:
Change locking in `removeSensor` to perform each sensor/metric removal atomically

junrao commented:
Thanks for the latest patch. LGTM.

auradkar commented:
Thanks Ismael. I think your latest patch should address my comments.
